src/keyrefresh/participant.rs

```diff
@@ -0,0 +1,1065 @@
+        // Without it we'll get a `InternalInvariantFailed` error when trying to
+        // retrieve `Decommit` below.
+        if !self.local_storage.contains::<storage::Decommit>(self.id()) {
+            let more_messages = run_only_once!(self.gen_round_one_msgs(rng, sid))?;
+            messages.extend_from_slice(&more_messages);
+        }
+
+        let decom = self
+            .local_storage
+            .retrieve::<storage::Decommit>(self.id())?;
+        let more_messages = self.message_for_other_participants(
+            MessageType::Keyrefresh(KeyrefreshMessageType::R2Decommit),
+            decom,
+        )?;
+        messages.extend_from_slice(&more_messages);
+        Ok(messages)
+    }
+
+    /// Handle the protocol's round two messages.
+    ///
+    /// Here we check that the decommitments from each participant are valid.
+    #[cfg_attr(feature = "flame_it", flame("keyrefresh"))]
+    #[instrument(skip_all, err(Debug))]
+    fn handle_round_two_msg<R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        message: &Message,
+    ) -> Result<ProcessOutcome<<Self as ProtocolParticipant>::Output>> {
+        info!("Handling round two keyrefresh message.");
+        // We must receive all commitments in round 1 before we start processing
+        // decommits in round 2.
+        let r1_done = self
+            .local_storage
+            .contains_for_all_ids::<storage::Commit>(&self.all_participants());
+        if !r1_done {
+            // Store any early round 2 messages
+            self.stash_message(message)?;
+            return Ok(ProcessOutcome::Incomplete);
+        }
+        // Check that the decommitment contained in the message is valid against
+        // the previously received commitment and protocol rules.
+        let decom = KeyrefreshDecommit::from_message(message)?;
+        let com = self
+            .local_storage
+            .retrieve::<storage::Commit>(message.from())?;
+        decom.verify(
+            message.id(), // Actually sid.
+            message.from(),
+            com,
+            &self.all_participants(),
+        )?;
+        self.local_storage
+            .store::<storage::Decommit>(message.from(), decom);
+
+        // Check if we've received all the decommits
+        let r2_done = self
+            .local_storage
+            .contains_for_all_ids::<storage::Decommit>(&self.all_participants());
+
+        if r2_done {
+            // Generate messages for round 3...
+            let round_three_messages = run_only_once!(self.gen_round_three_msgs(rng))?;
+
+            // ...and handle any messages that other participants have sent for round 3.
+            let mut round_three_outcomes = self
+                .fetch_messages(MessageType::Keyrefresh(KeyrefreshMessageType::R3Proofs))?
+                .iter()
+                .map(|msg| self.handle_round_three_msg(msg))
+                .collect::<Result<Vec<_>>>()?;
+
+            round_three_outcomes.extend(
```

Comment:
There is probably too much going on in this function call. Can we introduce a variable to bind everything inside the `extend(...)`?

Reply:
Done.